src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -2984,6 +2984,7 @@ ForInitResult analyzeForInit() {
         int depth = 0;
         boolean inType = false;
         boolean inSelectionAndParenthesis = false;
+        boolean sequentialIdentifiers = false;
```

Comment:
In general the changes look good - but please make sure that we don't get confused by lambda expressions. E.g. a lambda can be like `(type name)`, so you have two identifiers between parens. Or, even `(int x)`, so `int x` between parens - which I think will trick this code into thinking you have a record?

Reply:
This qualifies as a test case to add and a new bug to fix :D 
Good catch.

```
static <T> void method2(Function<Integer, Integer> f) {}
int i = 42;
for (method2((Integer a) -> 42); i == 0;) { i++; }
```

Reply:
Note also that a generic type might end with `>` - so `>` followed by an IDENT could also be a sign or a record (same as IDENT IDENT).

Reply:
PR updated!